File: TopCoder/src/main/java/com/leetcode/BinaryTreeMaximumPathSum.java
package com.leetcode;

import com.leetcode.tree.TreeNode;

//https://leetcode.com/problems/binary-tree-maximum-path-sum/
public class BinaryTreeMaximumPathSum {
    private int maxSum = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        recurisve(root);
        return maxSum;
    }

    private int recurisve(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int lSum = recurisve(root.left);
        int rSum = recurisve(root.right);
        int totalValue = root.val + lSum+ rSum;
        if (totalValue < 0) {
            maxSum = Math.max(maxSum, totalValue);
            return 0;
        }
        maxSum = Math.max(maxSum, totalValue);
        return root.val + Math.max(lSum, rSum);
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(2);
        root.left = new TreeNode(-1);

        int result = new BinaryTreeMaximumPathSum().maxPathSum(root);
        System.out.println(result);
    }
}

File: TopCoder/python/misc/BinaryTreeMaximumPathSum.py
class TreeNode:
    def __init__(self, value):
        self.val = value
        self.left = None
        self.right = None

class Solution:
    max_value = float('-inf')
    def maxPathSum(self, root):
        self.recursive(root)
        return self.max_value

    def recursive(self, root):
        if not root:
            return 0

        l_sum = self.recursive(root.left)
        r_sum = self.recursive(root.right)

        total_value = root.val + l_sum + r_sum

        if total_value < 0:
            self.max_value = max(self.max_value, total_value)
            return 0

        self.max_value = max(self.max_value, total_value)
        return root.val + max(l_sum, r_sum)

if __name__ == '__main__':
    root = TreeNode(2)
    root.left = TreeNode(1)
    root.right = TreeNode(3)

    sol = Solution()
    print(sol.maxPathSum(root))